kart/raster/pam_util.py

```diff
@@ -0,0 +1,53 @@
+import re
+
+
+def is_same_xml_ignoring_stats(lhs, rhs):
+    """
+    Returns True if two strings (left-hand-side, right-hand-side) of XML are the same,
+    or if they are the same except that gdal has inserted statistics into one of them,
+    in the form of a <Histograms> and/or <Metadata> block.
+    """
+    return (
+        lhs == rhs
+        or _lhs_is_rhs_minus_stats(lhs, rhs)
+        or _lhs_is_rhs_minus_stats(rhs, lhs)
+    )
+
+
+ANY_TAG = r"(<[^<>]+>)"
+HISTOGRAMS_PATTERN = re.compile(
+    rf"{ANY_TAG}\s*<Histograms>(.*)</Histograms>\s*{ANY_TAG}", re.DOTALL
+)
+# TODO - maybe check for non-statistics metadata.
+METADATA_PATTERN = re.compile(
+    rf"{ANY_TAG}\s*<Metadata>(.*)</Metadata>\s*{ANY_TAG}", re.DOTALL
+)
+
+
+def _lhs_is_rhs_minus_stats(lhs, rhs):
+    # Regex is sufficient for doing what we need to do here, which saves us parsing the XML.
```

Comment:
not a big fan of this - any reason to not parse the XML?

I can see one issue with the regex which is where there are multiple Metadata or Histogram tags with something else in between - the regex will greedily match the whole lot and this function will return True.

That specific issue could be fixed by adding a `?` in the right place to make it match non-greedily but it seems like it'd be cleaner to parse the XML

Reply:
Yeah you're right - this function doesn't work properly when there are multiple bands, so I need to rework it.

Basically using regex was the most straightforward and also efficient way of doing `==` comparisons on large chunks of XML. The XML libraries we're using so far don't provide an easy way to see if an XML element in one document is equivalent to another XML element in the other document. I can write such a comparison operator but it seemed a bit circuitous - first we parse the XML into a tree, then we flatten the tree back into one of (a string, some bytes, a list of strings) so we can compare it to another flattened tree. If I leave it flat (as the regex does) I've saved a few steps and it will run faster... might be on the wrong track though